src/viadot/sources/salesforce.py

```diff
@@ -139,3 +140,66 @@ def to_df(
             self.logger.info("Successfully downloaded data from the Salesforce API.")
 
         return df
+
+    def upsert(
+        self,
+        df: pd.DataFrame,
+        table: str,
+        external_id: str | None = None,
+        raise_on_error: bool = False,
+    ) -> None:
+        """Upsert data into Salesforce.
+
+        Args:
+            df (pd.DataFrame): Selected rows from Salesforce.
+            table (str): Table name.
+            external_id (str, optional): External ID. Defaults to None.
+            raise_on_error (bool, optional): Whether to raise on error.
+                Defaults to False.
+
+        Returns:
+            None
+        """
+        if df.empty:
+            self.logger.warning("No data to upsert.")
+            return
```

Comment:
Consider adding `self._handle_if_empty(if_empty)` - from the Base class. It's used in many sources.

Reply:
done